src/System.Management.Automation/engine/SessionStateLocationAPIs.cs

```diff
@@ -590,6 +590,79 @@ internal PathInfo SetLocation(string path, CmdletProviderContext context, bool l
             return this.CurrentLocation;
         }
 
+        /// <summary>
+        /// Changes the current working directory to the path specified.
+        /// It is only for FileSystem provider paths and optimized to be fast.
+        /// </summary>
+        /// <param name="path">
+        /// The file system path of the new current working directory.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        /// If <paramref name="path"/> is null.
+        /// </exception>
+        /// <exception cref="DriveNotFoundException">
+        /// If <paramref name="path"/> refers to a drive that does not exist.
+        /// </exception>
+        internal void SetLocationFileSystemFast(string path)
+        {
+            if (path == null)
+            {
+                throw PSTraceSource.NewArgumentNullException(nameof(path));
+            }
+
+            PathInfo current = null;
+            if (PublicSessionState.InvokeCommand.LocationChangedAction != null)
+            {
+                // It is under a condition to avoid the unneeded allocation - CurrentLocation always allocates.
+                current = CurrentLocation;
+            }
+
+            PSDriveInfo previousWorkingDrive = CurrentDrive;
+
+            if (LocationGlobber.IsProviderDirectPath(path))
+            {
+                // The path is a provider-direct path so use the current
+                // provider and its hidden drive but don't modify the path
+                // at all.
+                ProviderInfo provider = CurrentLocation.Provider;
+                CurrentDrive = provider.HiddenDrive;
+            }
+            else
+            {
+                // See if the path is a relative or absolute
+                // path.
+                if (Globber.IsAbsolutePath(path, out string driveName))
```

Comment:
We know this is an absolute path?  Perhaps we should verify in case someone calls this not understanding the assumptions.

Reply:
We call the method twice with:
- Directory.GetCurrentDirectory()
- System.IO.Path.GetDirectoryName(PsUtils.GetMainModule(currentProcess).FileName)
The both is absolute file system paths on Windows so that we have no need to reassign current drive - it is the same.

I added a commit with the cleanup.